File: app/src/main/java/com/one4ll/xplayer/MainActivity.java
package com.one4ll.xplayer;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import com.google.android.exoplayer2.ExoPlaybackException;
import com.google.android.exoplayer2.PlaybackParameters;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.Timeline;
import com.google.android.exoplayer2.audio.AudioAttributes;
import com.google.android.exoplayer2.audio.AudioListener;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.ProgressiveMediaSource;
import com.google.android.exoplayer2.source.TrackGroupArray;
import com.google.android.exoplayer2.trackselection.TrackSelectionArray;
import com.google.android.exoplayer2.ui.PlayerView;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
import com.google.android.exoplayer2.util.Util;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";
    @BindView(R.id.main_activity_constraint_layout)
    ConstraintLayout constraintLayout;
    @BindView(R.id.player_view)
    PlayerView playerView;
//    @BindView(R.id.progressBar)
//    ProgressBar progressBar;
//    @BindView(R.id.image_button_play)
//    ImageButton playButton;
//    @BindView(R.id.image_button_pause)
//    ImageButton pauseButton;
//    @BindView(R.id.image_button_play_2)
//    ImageButton playButton2;
//    @BindView(R.id.image_button_pause_2)
//    ImageButton pauseButton2;
//    @BindView(R.id.imageButtonPrevious)
//    ImageButton imageButtonPrevious;
//    @BindView(R.id.imageButtonNext)
//    ImageButton imageButtonNext;

    private SimpleExoPlayer simpleExoPlayer;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        simpleExoPlayer = new SimpleExoPlayer.Builder(this).build();
        playerView.setPlayer(simpleExoPlayer);
        Intent intent = getIntent();
        String videoUriPath = intent.getStringExtra("video");

//        final Uri fileUri = Uri.parse("android.resource://" + getPackageName() + "/" + R.raw.video);
//        Uri fileUri2 = Uri.parse("https://www.videvo.net/videvo_files/converted/2018_01/preview/171124_H1_005.mp436952.webm");
//        Uri fileUri3 = Uri.parse("http://185.105.103.101/serial/Supernatural/S13/720p.x265/Supernatural.S13E07.720p.HDTV.2CH.x265.HEVC.Filmaneh.mkv");

        Log.d(TAG, "onCreate: app name " + getString(R.string.app_name));
        DataSource.Factory dataSourceFactory = new DefaultDataSourceFactory(MainActivity.this, Util.getUserAgent(MainActivity.this, getString(R.string.app_name)));
        MediaSource mediaSource = new ProgressiveMediaSource.Factory(dataSourceFactory).createMediaSource(Uri.parse(videoUriPath));
        simpleExoPlayer.prepare(mediaSource);
        simpleExoPlayer.addAudioListener(audioListener);
        simpleExoPlayer.setPlayWhenReady(true);
        simpleExoPlayer.addListener(eventListener);
         LocalBroadcastManager.getInstance(this).registerReceiver(broadcastReceiver,new IntentFilter("customIntent"));

    }
    private  BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            Log.d(TAG, "onReceive: " + intent.getStringExtra("message"));
        }
    };

//    @OnClick(R.id.image_button_play)
//    public void play() {
//        playVideo();
//    }

//    @OnClick(R.id.image_button_play_2)
//    public void play2() {
//        playVideo();
//    }

//    private void playVideo() {
//        Log.d(TAG, "play: clicked");
//        if (simpleExoPlayer.getPlaybackState() == Player.STATE_ENDED) {
//            simpleExoPlayer.seekTo(0);
//        }
//        simpleExoPlayer.setPlayWhenReady(true);
//        playButton.setVisibility(View.INVISIBLE);
//        playButton2.setVisibility(View.INVISIBLE);
//        pauseButton2.setVisibility(View.VISIBLE);
//
//    }


//    @OnClick(R.id.player_view)
//    public void onPlayViewClicked() {
//        pauseButton.setVisibility(View.VISIBLE);
//        pauseButton2.setVisibility(View.VISIBLE);
//    }

//    @OnClick(R.id.image_button_pause)
//    public void pause() {
//        pauseVideo();
//    }

//    @OnClick(R.id.image_button_pause_2)
//    public void pause2() {
//        pauseVideo();
//    }

//    @OnClick(R.id.imageButtonPrevious)
//    public void goToPrevious() {
//        int state = simpleExoPlayer.getPlaybackState();
//    }

//    private void pauseVideo() {
//        progressBar.setVisibility(View.GONE);
//        simpleExoPlayer.setPlayWhenReady(false);
//        pauseButton.setVisibility(View.INVISIBLE);
//        pauseButton2.setVisibility(View.INVISIBLE);
//        playButton.setVisibility(View.VISIBLE);
//        playButton2.setVisibility(View.VISIBLE);
//
//    }


    private Player.EventListener eventListener = new Player.EventListener() {
        @Override
        public void onTimelineChanged(Timeline timeline, int reason) {
            Log.d(TAG, "onTimelineChanged: timeline " + timeline.toString());
            Log.d(TAG, "onTimelineChanged: reason " + reason);
        }

        @Override
        public void onTimelineChanged(Timeline timeline, @Nullable Object manifest, int reason) {
            Log.d(TAG, "onTimelineChanged: timeline " + timeline);
            Log.d(TAG, "onTimelineChanged: reason " + reason);

        }

        @Override
        public void onTracksChanged(TrackGroupArray trackGroups, TrackSelectionArray trackSelections) {

        }

        @Override
        public void onLoadingChanged(boolean isLoading) {
//            if (isLoading) {
//                progressBar.setVisibility(View.VISIBLE);
//            } else {
//                progressBar.setVisibility(View.GONE);
//            }

        }

        @Override
        public void onPlayerStateChanged(boolean playWhenReady, int playbackState) {
//            switch (playbackState) {
//                case Player.STATE_READY:
//                    Log.d(TAG, "onPlayerStateChanged: state ready");
//                    progressBar.setVisibility(View.INVISIBLE);
//                    break;
//                case Player.STATE_BUFFERING:
//                    progressBar.setVisibility(View.VISIBLE);
//                    Log.d(TAG, "onPlayerStateChanged: state buffering");
//                    break;
//                case Player.STATE_ENDED:
//                    Log.d(TAG, "onPlayerStateChanged: state Ended");
//                    playButton.setVisibility(View.VISIBLE);
//                    break;
//
//            }

        }

        @Override
        public void onPlaybackSuppressionReasonChanged(int playbackSuppressionReason) {

        }

        @Override
        public void onIsPlayingChanged(boolean isPlaying) {
            Log.d(TAG, "onIsPlayingChanged: isPlaying" + isPlaying);
//            if (isPlaying) {
//                progressBar.setVisibility(View.INVISIBLE);
//            }

        }

        @Override
        public void onRepeatModeChanged(int repeatMode) {

        }

        @Override
        public void onShuffleModeEnabledChanged(boolean shuffleModeEnabled) {

        }

        @Override
        public void onPlayerError(ExoPlaybackException error) {
            error.printStackTrace();
            Log.d(TAG, "onPlayerError:" + error.getMessage());
            Toast.makeText(MainActivity.this, error.getMessage(), Toast.LENGTH_LONG).show();

        }

        @Override
        public void onPositionDiscontinuity(int reason) {

        }

        @Override
        public void onPlaybackParametersChanged(PlaybackParameters playbackParameters) {

        }

        @Override
        public void onSeekProcessed() {

        }
    };
    private AudioListener audioListener = new AudioListener() {
        @Override
        public void onAudioSessionId(int audioSessionId) {
            Log.d(TAG, "onAudioSessionId: " + audioSessionId);
        }

        @Override
        public void onAudioAttributesChanged(AudioAttributes audioAttributes) {
            Log.d(TAG, "onAudioAttributesChanged: audioAttributes content type" + audioAttributes.contentType);
            Log.d(TAG, "onAudioAttributesChanged: audioAttributes flags" + audioAttributes.flags);
            Log.d(TAG, "onAudioAttributesChanged: audioAttributes usage" + audioAttributes.usage);
            Log.d(TAG, "onAudioAttributesChanged: audioAttributes allowed capture policy" + audioAttributes.allowedCapturePolicy);
        }

        @Override
        public void onVolumeChanged(float volume) {
            Log.d(TAG, "onVolumeChanged: " + volume);
        }
    };

    @Override
    protected void onDestroy() {
        simpleExoPlayer.release();
        super.onDestroy();
        LocalBroadcastManager.getInstance(this).unregisterReceiver(broadcastReceiver);
    }

    public String convertDuration(long duration){
        String out = null;
        long hours = 0;
        try {
            hours = (duration / 3600000);
        }catch (Exception e ){
            e.printStackTrace();
        }
        long remainingMinutes = (duration - (hours * 3600000)) / 60000;
        String minutes = String.valueOf(remainingMinutes);
        if (minutes.equals("0")) {
            minutes = "00";
        }
        long remainingSecs = (duration - (hours * 3600000))  - (remainingMinutes * 60000);
        String seconds = String.valueOf(remainingSecs);
        if (seconds.length() < 2){
            seconds = "00";
        }else {
            seconds = seconds.substring(0, 2);
        }
        if (hours > 0){
            out = hours +":" + minutes+ ":" + seconds;
        }else {
            out = minutes + ":" + seconds;
        }
       return out;
    }

}

File: app/src/main/java/com/one4ll/xplayer/MainActivity.kt
package com.one4ll.xplayer

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.media.AudioManager
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.util.DisplayMetrics
import android.util.Log
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.ui.PlayerView
import com.google.android.exoplayer2.upstream.DataSource
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory
import com.google.android.exoplayer2.util.Util
import com.one4ll.xplayer.custom.CustomPlayerEventListener
import com.one4ll.xplayer.helpers.*
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

private const val TAG = "MainActivity"

@AndroidEntryPoint
class MainActivity : AppCompatActivity(), View.OnTouchListener, GestureDetector.OnDoubleTapListener,
    GestureDetector.OnGestureListener, AudioManager.OnAudioFocusChangeListener {
    private lateinit var gestureDetector: GestureDetector
    private lateinit var videoUriPath: String
    private lateinit var playerView: PlayerView
    private lateinit var goToTextView: TextView
    private lateinit var crossImageView: ImageView
    private lateinit var videoTitle: TextView
    private var screenWidth: Int? = null
    private var screenHeight: Int? = null
    private val brightness by lazy {
        Settings.System.getInt(
            contentResolver,
            Settings.System.SCREEN_BRIGHTNESS
        )
    }

    @Inject
    lateinit var audioManager: AudioManager

    @Inject
    lateinit var simpleExoPlayer: SimpleExoPlayer
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        playerView = findViewById(R.id.player_view)
        goToTextView = findViewById(R.id.goto_duration)
        crossImageView = findViewById(R.id.cross_image)
        videoTitle = findViewById(R.id.video_title)
        hideSystemUI(true)
        val displayMetrics = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(displayMetrics)
        screenWidth = displayMetrics.widthPixels
        screenHeight = displayMetrics.heightPixels
        Log.d(TAG, "onCreate: screen width $screenWidth ,screen  height $screenHeight")
        gestureDetector = GestureDetector(this, this)
        playerView.player = simpleExoPlayer
        if (intent.action != null && intent.action == Intent.ACTION_VIEW) {
            Log.d(TAG, "onCreate: action view intent called")
            videoUriPath = intent.data.toString()
            Log.d(TAG, "onCreate: action view intent called $videoUriPath")
        } else {
            videoUriPath = intent.getStringExtra(VIDEO_PATH)
                ?: throw IllegalArgumentException(getString(R.string.videoPathShouldnotNull))
        }
        setVideoTitle()
        setViewListener()

        val dataSourceFactory: DataSource.Factory = DefaultDataSourceFactory(
            this,
            Util.getUserAgent(this@MainActivity, getString(R.string.app_name))
        )
        Log.d(TAG, "onCreate: Video Path $videoUriPath")
        lateinit var mediaSource: MediaSource
        mediaSource =
            ProgressiveMediaSource.Factory(dataSourceFactory)
                .createMediaSource(Uri.parse(videoUriPath))
        simpleExoPlayer.prepare(mediaSource)

        simpleExoPlayer.playWhenReady = true
        val eventListener = CustomPlayerEventListener(
            goToTextView = goToTextView,
            playerView = playerView,
            simpleExoPlayer = simpleExoPlayer
        )
        simpleExoPlayer.addListener(eventListener)

    }

    private fun setViewListener() {
        playerView.setOnTouchListener(this)
        crossImageView.setOnClickListener { finish() }
    }

    private fun setVideoTitle() {
        videoTitle.text = videoUriPath.getTitleFromVideoPath()
    }


    override fun onAudioFocusChange(focusChange: Int) {
        when (focusChange) {
            AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_MAY_DUCK -> {
                simpleExoPlayer.playWhenReady = true
            }
            AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK -> {
                simpleExoPlayer.playWhenReady = false
            }
            AudioManager.AUDIOFOCUS_LOSS -> {
                simpleExoPlayer.playWhenReady = false
            }
        }
    }


    override fun onTouch(view: View?, event: MotionEvent?): Boolean {
        view?.performClick()
        gestureDetector.onTouchEvent(event)
        playerView.showController()
        return true
    }

    override fun onDoubleTap(event: MotionEvent?): Boolean {
        forwardAndBackWardVideoOnDoubleClick(event)
        return true
    }

    private fun forwardAndBackWardVideoOnDoubleClick(event: MotionEvent?) {
        Log.d(TAG, "onDoubleTap: clicked")
        val rawX = event?.rawX
        Log.d(TAG, "forwardAndBackWardVideoOnDoubleClick: raw x $rawX")
        if (rawX != null) {
            val middle = screenWidth!! / 2
            if (rawX < middle) {
                Log.d(TAG, "forwardAndBackWardVideoOnDoubleClick: Left side")
                val currentPosition = simpleExoPlayer.currentPosition
                simpleExoPlayer.seekTo(currentPosition - SKIP_DURATION)
            } else {
                Log.d(TAG, "forwardAndBackWardVideoOnDoubleClick: Right side")
                val currentPosition = simpleExoPlayer.currentPosition
                simpleExoPlayer.seekTo(currentPosition + SKIP_DURATION)

            }
        }
    }

    override fun onDoubleTapEvent(e: MotionEvent?): Boolean {
        Log.d(TAG, "onDoubleTapEvent: ")
        return true
    }

    override fun onSingleTapConfirmed(e: MotionEvent?): Boolean {
        return true

    }

    override fun onShowPress(e: MotionEvent?) {
    }

    override fun onSingleTapUp(e: MotionEvent?): Boolean {
        playerView.controllerAutoShow =
            !playerView.controllerAutoShow
        return true
    }

    override fun onDown(e: MotionEvent?): Boolean {
        return true
    }

    override fun onFling(
        e1: MotionEvent?,
        e2: MotionEvent?,
        velocityX: Float,
        velocityY: Float
    ): Boolean {
        Log.d(TAG, "onFling: velocity x $velocityX velocity y $velocityY")
        return true
    }

    @SuppressLint("SetTextI18n")
    override fun onScroll(
        motionEvent1: MotionEvent?,
        motionEvent2: MotionEvent?,
        distanceX: Float,
        distanceY: Float
    ): Boolean {
        Log.d(TAG, "onScroll: motion event 1$motionEvent1")
        Log.d(TAG, "onScroll: motion event 2 $motionEvent2")
        goToTextView.visibility = View.VISIBLE
        val currentPosition = simpleExoPlayer?.currentPosition ?: 0
        if (motionEvent2?.x!! - motionEvent1?.x!! >= 100) {
            //forward
            goToTextView.text =
                "Forward  +20.00s\n\t\t ${convertDuration(currentPosition)}s"
            simpleExoPlayer.seekTo(currentPosition + 20000)

        } else if (motionEvent1.x - motionEvent2.x >= 100) {
            //backward
            goToTextView.text =
                "BackWard -20.00s\n\t\t ${convertDuration(currentPosition)}"
            simpleExoPlayer.seekTo(currentPosition - 20000)
        } else {
            val rawX = motionEvent1.rawX
            val middle = screenWidth!! / 2
            if (rawX < middle) {
                //left brightness
                if (motionEvent1.y - motionEvent2.y >= 100) {
                    //increase brightness

                    increaseBrightness()

                } else if (motionEvent2.y - motionEvent1.y >= 100) {
                    // decrease brightness
                    decreaseBrightness()
                }
            } else {
                //right sound
                if (motionEvent1.y - motionEvent2.y > 100) {
                    //increase sound
                    increaseSound()
                } else if (motionEvent2.y - motionEvent1.y >= 100) {
                    //decrease sound
                    decreaseSound()
                }
                //change sound
            }
        }
        Log.d(TAG, "onScroll: distance x $distanceX and distant y $distanceY")
        return true
    }

    private fun decreaseSound() {
        audioManager.adjustVolume(AudioManager.ADJUST_LOWER, AudioManager.FLAG_PLAY_SOUND)
    }

    private fun increaseSound() {
        Log.d(TAG, "increaseSound:")
        audioManager.adjustVolume(AudioManager.ADJUST_RAISE, AudioManager.FLAG_PLAY_SOUND)
    }

    // TODO: Add brightness control
    private fun decreaseBrightness() {
        Log.d(TAG, "decreaseBrightness:")
    }

    private fun increaseBrightness() {
        Log.d(TAG, "increaseBrightness: brightness $brightness")
        if (havePermission(this, Manifest.permission.WRITE_SETTINGS)) {

            Settings.System.putInt(contentResolver, Settings.System.SCREEN_BRIGHTNESS, 23)
            val attributes = window.attributes
            attributes.screenBrightness = (23 / 255).toFloat()
            window.attributes = attributes
        }
    }

    override fun onLongPress(e: MotionEvent?) {

    }

    override fun onPause() {
        super.onPause()
        simpleExoPlayer.playWhenReady = false
    }
}
